### PR TITLE
Add investigation document for multi-cluster support in Capact

### PR DESCRIPTION
**Pull request by mszostok**
## Description

Changes proposed in this pull request:

- Add investigation document for multi-cluster support in Capact

## Notes

_No need to check that._
- rejected: https://gist.github.com/mszostok/7817cc1f43c922c04f1120bde1542379

### CODE REVIEW

File: docs/investigation/multi-cluster/README.md [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+      oneOf:
+        - name: kubernetes
+          alias: k8s
+          revision: 0.1.0
+
+# ...
+            steps:
+              - - name: helm-install
+                  capact-action: helm.install
+                  capact-target: k8s # based on alias in requires section
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/README.md:123]: So effectively `capact-target` would modify the step, adding `capact-policy` property with policy injecting a given Kubeconfig (the one preregistered in cluster or provided as `capact-target`), correcto?

File: docs/investigation/multi-cluster/README.md [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+
+### Consequences
+
+To reduce the boilerplate and support multi-cluster in Capact, the following items need to be resolved:
+
+1. Remove `cap.attribute.containerization.kubernetes.kubeconfig-support` from **attributes**.
+2. Merge the `cap.type.containerization.kubernetes.kubeconfig` to `cap.core.type.platform.kubernets` as we don't have the [Type composition](https://capact.io/docs/feature/type-features#type-composition) yet.
+3. Allow injecting `cap.core.type.platform.kubernets` via Policy. We already use such approach for AWS and GCP credentials.
+4. Add option to use `inject.requiredTypeInstances` via artifact name in Workflow Policy. This will solve https://github.com/capactio/capact/issues/538 as we will do that via Policy instead.
+5. Solve [Support setting relations for optional TypeInstances in workflows](https://github.com/capactio/capact/issues/537) issue, but also take into account TypeInstances from the `requires` section.
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/README.md:151]: > but also take into account TypeInstances from the `requires` section.

Isn't that duplication of the point No 4 in consequences?

File: docs/investigation/multi-cluster/admiralty.md [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+9. Label the default Namespace in the Capact cluster to enable multi-cluster scheduling at the namespace level:
+
+    ```bash
+    kubectl --context kind-dev-capact label ns default multicluster-scheduler=enabled
+    ```
+
+## Demo
+
+1. Create Argo Workflow in Capact cluster, targeting the `eu` workload cluster:
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/admiralty.md:141]:
```suggestion
1. Create Job in Capact cluster, targeting the `eu` workload cluster:
```

File: docs/investigation/multi-cluster/admiralty.md [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+
+    ```bash
+    argo watch eu-foo
+    ```
+
+3. Check Argo Workflow logs:
+
+    ```bash
+    argo log eu-foo
+    ```
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/admiralty.md:179]: tenor-13023104

File: docs/investigation/multi-cluster/admiralty.md [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    spec:
+      kubeconfigSecret:
+        name: eu
+    EOF
+    ```
+
+7. In the workload clusters, create a Source for the Capact cluster:
+
+     ```yaml
+     cat <<EOF | kubectl --context kind-$CLUSTER_NAME apply -f -
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/admiralty.md:117]:
```suggestion
7. In the workload clusters, create a Source for the workload cluster:

     ```yaml
     cat <<EOF | kubectl --context kind-eu apply -f -
```

Right?

File: docs/investigation/multi-cluster/admiralty.md [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+3. Check Argo Workflow logs:
+
+    ```bash
+    argo log eu-foo
+    ```
+
+4. Check that the pod was scheduled on `eu` workload node:
+
+    ```bash
+    kubectl --context kind-eu get pods -o wide -n default
```
Comment (pkosiec) [inline docs/investigation/multi-cluster/admiralty.md:184]: FYI I couldn't make it able to work, but I didn't want to waste time for investigation, and also I was using k3d cluster (as the Capact cluster).

I get the idea though 🙂 

```
  ----     ------            ----   ----             -------
  Warning  FailedScheduling  7m17s  admiralty-proxy  0/2 nodes are available: 1 Insufficient cpu, 1 Too many pods, 1 node(s) didn't match node selector.
  Warning  FailedScheduling  7m17s  admiralty-proxy  0/2 nodes are available: 1 Insufficient cpu, 1 Too many pods, 1 node(s) didn't match node selector.
```

```
❯ k describe nodes admiralty-default-eu
Name:               admiralty-default-eu
Roles:              cluster
Labels:             alpha.service-controller.kubernetes.io/exclude-balancer=true
                    kubernetes.io/role=cluster
                    type=virtual-kubelet
                    virtual-kubelet.io/provider=admiralty
Annotations:        node.alpha.kubernetes.io/ttl: 0
CreationTimestamp:  Thu, 25 Nov 2021 14:33:34 +0100
Taints:             virtual-kubelet.io/provider=admiralty:NoSchedule
Unschedulable:      false
Lease:
  HolderIdentity:  admiralty-default-eu
  AcquireTime:     <unset>
  RenewTime:       Thu, 25 Nov 2021 14:43:34 +0100
Conditions:
  Type    Status  LastHeartbeatTime                 LastTransitionTime                Reason  Message
  ----    ------  -----------------                 ------------------                ------  -------
  Ready   True    Thu, 25 Nov 2021 14:43:34 +0100   Thu, 25 Nov 2021 14:33:34 +0100
Addresses:
  InternalIP:  10.42.0.124
System Info:
  Machine ID:
  System UUID:
  Boot ID:
  Kernel Version:
  OS Image:
  Operating System:
  Architecture:
  Container Runtime Version:
  Kubelet Version:
  Kube-Proxy Version:
PodCIDR:                      10.42.1.0/24
PodCIDRs:                     10.42.1.0/24
Non-terminated Pods:          (0 in total)
  Namespace                   Name    CPU Requests  CPU Limits  Memory Requests  Memory Limits  Age
  ---------                   ----    ------------  ----------  ---------------  -------------  ---
Allocated resources:
  (Total limits may be over 100 percent, i.e., overcommitted.)
  Resource           Requests  Limits
  --------           --------  ------
  cpu                0 (0%)    0 (0%)
  memory             0 (0%)    0 (0%)
  ephemeral-storage  0 (0%)    0 (0%)
Events:              <none>
```

File: docs/investigation/multi-cluster/README.md (hunk 2 above)
Comment (mszostok) [reply docs/investigation/multi-cluster/README.md:151]: discussed f2f, invalid 👍

**Review by pkosiec — APPROVED** [review]
We already discussed f2f the whole concept and I agree that the "Capact provisions Capact" option is the best one currently. Apart from some minor comments, it looks good to me 👍

File: docs/investigation/multi-cluster/README.md (hunk 1 above)
Comment (mszostok) [reply docs/investigation/multi-cluster/README.md:123]: discussed f2f

**Pull request merged**